internal/serialization/class_definition_context.go

```diff
@@ -28,32 +29,39 @@ import (
 
 type ClassDefinitionContext struct {
 	classDefs map[string]*serialization.ClassDefinition
+	mu        *sync.RWMutex
 	factoryID int32
 }
 
 func NewClassDefinitionContext(factoryID int32) *ClassDefinitionContext {
 	return &ClassDefinitionContext{
 		factoryID: factoryID,
 		classDefs: make(map[string]*serialization.ClassDefinition),
+		mu:        &sync.RWMutex{},
 	}
 }
 
 func (c *ClassDefinitionContext) LookUp(classID int32, version int32) *serialization.ClassDefinition {
-	return c.classDefs[encodeVersionedClassID(classID, version)]
+	c.mu.RLock()
+	cd := c.classDefs[encodeVersionedClassID(classID, version)]
+	c.mu.RUnlock()
+	return cd
 }
 
-func (c *ClassDefinitionContext) Register(classDefinition *serialization.ClassDefinition) error {
+func (c *ClassDefinitionContext) register(classDefinition *serialization.ClassDefinition) error {
 	if classDefinition == nil {
 		return nil
 	}
 	if classDefinition.FactoryID != c.factoryID {
 		text := fmt.Sprintf("this factory's id is %d, intended factory id is %d.", c.factoryID, classDefinition.FactoryID)
 		return ihzerrors.NewSerializationError(text, nil)
 	}
-	classDefKey := encodeVersionedClassID(classDefinition.ClassID, classDefinition.Version)
-	current := c.classDefs[classDefKey]
+	k := encodeVersionedClassID(classDefinition.ClassID, classDefinition.Version)
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	current := c.classDefs[k]
 	if current == nil {
-		c.classDefs[classDefKey] = classDefinition
+		c.classDefs[k] = classDefinition
 		return nil
 	}
 	if !reflect.DeepEqual(current, classDefinition) {
```

Comment:
I'm not sure if this comment is right but do we expect this `DeepEqual` function to take short time? Because the mutex is unlocked when function returns and it might block other requests to continue. Maybe we can unlock the mutex after setting the `classDefs`?

Reply:
The time `DeepEqual` takes isn't significant in this context, also `register` method is not in the hot path.